lib/shopify_transporter/pipeline/magento/order/discounts.rb

```diff
@@ -0,0 +1,103 @@
+# frozen_string_literal: true
+require 'shopify_transporter/pipeline/stage'
+require 'shopify_transporter/shopify'
+
+module ShopifyTransporter
+  module Pipeline
+    module Magento
+      module Order
+        class Discounts < Pipeline::Stage
+          def convert(input, record)
+            record.merge!(
+              {
+                discounts: discounts(input),
+              }.stringify_keys
+            )
+          end
+
+          private
+
+          def discounts(hash)
+            [
+              shipping_discount(hash),
+              fixed_amount_discount(hash),
+              percentage_discount(hash),
+            ].compact
+          end
+
+          def shipping_discount(hash)
+            shipping_discount_amount = fetch_value(hash, 'shipping_discount_amount')
+            shipping_amount = fetch_value(hash, 'shipping_amount')
+            if shipping_discount_amount > shipping_amount
+              {
+                code: discount_code(hash),
+                amount: shipping_discount_amount,
+                type: 'shipping',
+              }.stringify_keys
+            end
+          end
+
+          def fixed_amount_discount(hash)
+            discount_amount = discount_amount(hash)
+            discount_percentage = discount_percentage(hash)
+            if discount_amount > 0 && discount_percentage == 0
+              {
+                code: discount_code(hash),
+                amount: discount_amount,
+                type: 'fixed_amount',
+              }.stringify_keys
+            end
+          end
+
+          def percentage_discount(hash)
+            discount_amount = discount_amount(hash)
+            discount_percentage = discount_percentage(hash)
+
+            if discount_amount > 0 && discount_percentage != 0
+              {
+                code: discount_code(hash),
+                amount: discount_percentage,
+                type: 'percentage',
+              }.stringify_keys
+            end
+          end
+
+          def discount_percentage(hash)
+            percentage = 0
```

Comment:
IMO, any time you write a function that looks like this:

```
def foo(a, b, c, ...)
  percentage = 0
  ...
  percentage
end
```

that's a smell, and it should signal to you that maybe there's a better way to do the same thing.

In this case, I'm not sure what the best way forward is, but here's my first gut reaction on how to write this function:

```
def discount_percentage(hash)
  return 0 unless hash.dig('items', 'result', 'items', 'item').present?

  if line_items(hash).is_a?(Hash)
    fetch_value(line_items(hash), 'discount_percent')
  else
    discounts = line_items(hash).map do |line_item|
      fetch_value(line_item, 'discount_percent')
    end.uniq
    discount_applied_on_all_line_items?(discounts) ? discounts.first : 0
  end
end
```

Any reason this wouldn't work?

Reply:
Also, maybe people actually find this less clear -- I personally prefer it, but interested in other people's opinions as well.

Reply:
Also honestly I would prefer just checking `discounts.length == 1` rather than having `discount_applied_on_all_line_items?` which is really just an unnecessary layer of indirection. Again, not sure about how other people feel about this.

Reply:
@JordanVerasamy Could u elaborate why this code style is considered as smelly code? I don't think I completely follow.

Reply:
I extracted the check of `discounts.length == 1` into a separate method because the name of the method explains the reason why this check is necessary. I'm open to others' opinions on this matter.